Add authentication by client

melnikovio opened the pull request:
Hi. Please review the option of adding client_credentials authentication. User-based authentication isn't always feasible or optimal. I added a separate client for this to ensure that existing applications backward compatibility remains intact.

fschick commented:
Thank you. I think from an architectural point of view, authentication via the 'Client Credentials Flow' is the better approach than the existing.

Some points to discuss:
* The added client `KeycloakClientHttpClient.cs` uses the same class name as `KeycloakHttpClient.cs`. I think this isn't intended
* When a separate client is created, a separate `ApiClientFactory` should be created too
* Maybe it would be better to have both (and mayby upcomming flows) in one class like

```csharp
// TODO: Check, if an existing enum in .NET Framework exists
public enum AuthenticationFlow
{
    ClientCredentials,
    PasswordGrant
}

[Obsolete("Use constructor with explicit authentication flow parameter.")]
public KeycloakHttpClient(string authServerUrl, string userOrClientId, string passwordOrClientSecret)
    : this(...) { }

public KeycloakHttpClient(string authServerUrl, string userOrClientId, string passwordOrClientSecret, AuthenticationFlow authenticationFlow)
    : this(...) { }
```

melnikovio commented:
Hi. Thank you for the feedback.

1. Fixed

3. Interestingly, the first time I did it exactly that way with both flows, but I felt that considering backward compatibility, the class was starting to become complex. In my opinion, both options have their place, so I added another implementations of both flows to KeycloakHttpClient.cs. Let's compare and choose the final one?

melnikovio commented:
> Hi. Thank you for the feedback.
> 
> 1. Fixed
> 2. It's funny that the first time I did it exactly that way with both flows, but I felt that considering backward compatibility, the class was starting to become complex. In my opinion, both options have their place, so I added another implementations of both flows to KeycloakHttpClient.cs. Let's compare and choose the final one?

And I haven't choose option with obsolete yet, I just made the flow optional. Should I go ahead and mark it as obsolete? This would require copying all the methods. Is it worth it?

fschick commented:
It's a mess. Either
* The class needs to be duplicated
* The constructors needs to be duplicated
* The default flow points to the legacy one `PasswordGrant`

The downside of class duplication is the class name itself. The class name should be clear, such as `KeycloakHttpClientClientCredentialFlow` and `KeycloakHttpPasswordGrantFlow`, which breaks with backward compatibly.

If you're fine with this, I'd like to go with `KeycloakHttpClient` and the optional `AuthenticationFlow flow = AuthenticationFlow.PasswordGrant` parameter for this pull request. 

I need to think about a cleaner way afterwards. IMHO constructors with a lot of overloads are not clean code. If you have a good idea, it is welcome.

melnikovio commented:
How about sticking to 'keep it simple'? For example, look at how many flows the Golang client implements here - https://github.com/Nerzal/gocloak/blob/main/client.go.

What if, for backward compatibility, leave the current class KeycloakHttpClient untouched and mark it as obsolete, while creating a new one alongside it? In the new interface, we can pass the T<Flow Parameters>.
Then create separate classes for parameters. Additionally, can provide a simple Token class so that people like me could implement token retrieval themselves, as it's more convenient for them :)

public class PasswordGrantFlow
    {
        public string AuthUrl { get; set; }
        public string UserName { get; set; }
        public string Password { get; set;} 
    }

    public class ClientCredentialsFlow
    {
        public string AuthUrl { get; set; }
        public string ClientId { get; set; }
        public string ClientSecret { get; set; }
    }

    public class OfflineToken
    {
        public string AuthUrl { get; set; }
        public string Token { get; set; }
    }

fschick commented:
This is the right way, thank you. Would you like to implement it with this pull request or should I do the work (5-7 days until delivery)?

melnikovio commented:
I will continue what I started. I just need a little more time.

melnikovio commented:
Please take a look at the MR. It's very difficult to write tests or proper examples under such a structure...

fschick commented:
Thank you. I think the main work is done. The project has currently no tests (shame on me). It was a quick solution because I didn't find a complete C# client for Keycloak.

It seems you developed the code in a different solution/project with implicit imports, nullable reference types and latest C# features (`required` modifier) enabled. The project compiles currently against `netstandard2.0` to be as compatible as possible (e.g. .NET 4.8).

A solution for this repository can be generated via Powershell `Build/scripts/rest_api_client.publish.ps1` (ignore the errors about publishing NuGet packages). But you need an OpenJDK installed.

If you're fine with this, I will accept the pull request in the evening (UTC+1:00), fix the smaller things and create a release.

melnikovio commented:
Of course, no problem. I've removed everything new, and now it builds under netstandard 2.0. Also, updated readme with new methods.
The only thing is I won't be able to build it using the script since I'm not on Windows. Would it be possible for you to build and check it on your end to ensure everything is working correctly?

fschick commented:
It works now. Thank you.

Basically, PowerShell is available for Linux too. But not an optimla solution. I will try to switch to [NUKE Build](https://nuke.build/) for all my projects soon.
